SConstruct

```diff
@@ -98,73 +96,119 @@ for x in sorted(glob.glob("platform/*")):
-env_base.__class__.module_add_dependencies = methods.module_add_dependencies
-env_base.__class__.module_check_dependencies = methods.module_check_dependencies
+selected_platform = ""
 
-env_base["x86_libtheora_opt_gcc"] = False
-env_base["x86_libtheora_opt_vc"] = False
+if env_base["platform"] != "":
+    selected_platform = env_base["platform"]
+elif env_base["p"] != "":
+    selected_platform = env_base["p"]
+else:
+    # Missing `platform` argument, try to detect platform automatically
+    if (
+        sys.platform.startswith("linux")
+        or sys.platform.startswith("dragonfly")
+        or sys.platform.startswith("freebsd")
+        or sys.platform.startswith("netbsd")
+        or sys.platform.startswith("openbsd")
+    ):
+        selected_platform = "linuxbsd"
+    elif sys.platform == "darwin":
+        selected_platform = "macos"
+    elif sys.platform == "win32":
+        selected_platform = "windows"
+    else:
+        print("Could not detect platform automatically. Supported platforms:")
+        for x in platform_list:
+            print("\t" + x)
+        print("\nPlease run SCons again and select a valid platform: platform=<string>")
 
-# avoid issues when building with different versions of python out of the same directory
-env_base.SConsignFile(".sconsign{0}.dblite".format(pickle.HIGHEST_PROTOCOL))
+    if selected_platform != "":
+        print("Automatically detected platform: " + selected_platform)
 
-# Build options
+if selected_platform in ["macos", "osx"]:
+    if selected_platform == "osx":
+        # Deprecated alias kept for compatibility.
+        print('Platform "osx" has been renamed to "macos" in Godot 4. Building for platform "macos".')
+    # Alias for convenience.
+    selected_platform = "macos"
 
-customs = ["custom.py"]
+if selected_platform in ["ios", "iphone"]:
+    if selected_platform == "iphone":
+        # Deprecated alias kept for compatibility.
+        print('Platform "iphone" has been renamed to "ios" in Godot 4. Building for platform "ios".')
+    # Alias for convenience.
+    selected_platform = "ios"
 
-profile = ARGUMENTS.get("profile", "")
-if profile:
-    if os.path.isfile(profile):
-        customs.append(profile)
-    elif os.path.isfile(profile + ".py"):
-        customs.append(profile + ".py")
+if selected_platform in ["linux", "bsd", "x11"]:
+    if selected_platform == "x11":
+        # Deprecated alias kept for compatibility.
+        print('Platform "x11" has been renamed to "linuxbsd" in Godot 4. Building for platform "linuxbsd".')
+    # Alias for convenience.
+    selected_platform = "linuxbsd"
+
+# Platform specific flags.
+if selected_platform in platform_flags:
+    platform_dict = dict(platform_flags[selected_platform])
+    for option in opts.options:
+        # Use the platform flag value if it's not set in the environment already (i.e. the user hasn't overridden it).
+        if env_base[option.key] == option.default and option.key in platform_dict:
```

Comment:
If a user explicitly sets an option to the default value (via command line or `custom.py`, and the platform flags set it to a different value, the platform flags will "take priority" in this case, because this check will think the user did not set the value.

Some different approach is needed, in which the platform flags are calculated first, and passed as the default values when building the options, for every option.

Reply:
Changing the default values for the options depending on the selected platform is a design decision with slightly wider implications than what this PR is trying to accomplish, which is to fix bugs in the early option parsing.

Changing the default values means explaining to the user why the defaults are suddenly different; updating the documentation to explain why they can change; and it can potentially break existing user workflows that expect the default values to be the same regardless of the platform.

I'm not saying it's not a potentially useful change, mind you! There is some sense to deciding that the platform is the one defining the defaults, and therefore the defaults should be set by the platform, and not the other way around. So, I wouldn't mind opening a PR with a potential change to do that, and having a discussion about it, yeah.

In the meantime, for this particular issue you raised, we can handle that without changing the defaults, at least for these very early options that control the early build environment. The problem here is that we can't actually tell whether the user has set a value for an option - whatever that value is - because Scons doesn't differentiate between "value == default because the user hasn't set any value" and "value == default because the user set it to that". 

There are two potential ways to fix that, for this particular part of the initialization:
- We can use our own Variables implementation that DOES track the difference between an option that has a default implied value and an option with an explicit user set value, or

- We don't set any defaults on the options at this stage of the initialization, so if they're empty, that means the user hasn't set a value, and we can safely use the value that the platform sets. If neither user nor platform set anything, we can then use the "real" default value that we already use in the full options set later in the code. This won't affect our help text later on, because that comes from the full set of options with actual default values.

I'm going to do the second one, because that's easy and will solve our immediate issue. And it will actually put everything in place for us to talk about changing the default values on a per-platform basis in the future - it would be very easy to inject the values that are returned from the platform flags here into the full option set that we define later, if we want to do that (later, in another PR).

Reply:
I don't know what you mean by "default options per platform", and I wasn't talking about that anyway. I was talking about: If the platform sets a flag, and the user does too, the user's selection should take priority.

If you succeed in making this work using `None` as the default (and later override it with the real default), that's great. But, I think it might not work for options that go through conversion, e.g. boolean options, because it would try to convert `None` to boolean and fail. But there's only one way to find out.

If you do get it to work, note that the fix should also be applied to any use of `ARGUMENTS` and `get_cmdline_bool` below. i.e. they all need to be replaced with a check of "If the value is `None`, replace it with this value".

P.S. I did suggest re-implementing `Variables` in RocketChat, but Akien was opposed.

Reply:
You said

> Some different approach is needed, in which the platform flags are calculated first, and passed as the default values when building the options, for every option.

This implies that the default values for the Options change depending on the selected platform. That's what I was referring to when I said it was a larger design decision than just respecting the user flags over the platform flags.

I'm not changing any usages of `get_cmdline_bool` or arguments right now because I'm not touching any flags that would be overridden by those direct reads, not right now anyways. This is already a complicated change due to how many things `use_mingw` and `target` affect, I don't want to add additional unrelated changes on top of it.

Once this one is good to go, we can do more fixes for other flags that aren't beyond properly obeyed. Not all of them are being handled in the same way, so they all require specific fixes.

I've pushed some fixes to properly obey the user flags over the platform flags.

Reply:
The platform flags _already_ override the default flags. There's no change there, larger or otherwise. The difference is that they currently do it for non-default flags as well.

As it is, you already copied the code that handles `platform_flags` from the bottom to the top, changed the top version to check the default value (and now, to check vs empty string), while leaving the bottom version as-is. That means the initial check for platform settings respects `custom.py`, but the actual usage does not. Meaning that, in the worst case, the settings used for building can be different than the ones used to load the tools.

Also, the empty string solution feels like a step in the wrong direction. `None` would have been nice if it worked, but empty string feels too fragile. It's replacing a bad hack with a worse hack.

After thinking about it for a while, there _is_ a clean solution that doesn't require rewriting `Variable`. Since Python doesn't have access specifiers. As such, it's possible to retroactively change the default value of a variable by changing `opts.options[i].default` where `opts.options[i].key` is the option's name. The new default can then be applied by calling `opts.Update`. Notably, this can be done any number of times for the same environment. This way, several overrides for the default values can be applied, without overriding user values.

While it may seem that a complete fix to the problem is out of scope for this PR, a partial fix that is so hacky it would need to be reverted later, is better off not being merged. Even if it does need to be done in two PRs, the PR fixing the default values issue has to come first, and this one has to be based on top of it. Otherwise, a large part of this PR will simply have to be reverted later on.